### PR TITLE
Fix TotalRewardPerEpisode bug

### PR DESCRIPTION
**Pull request by jeremiahpslewis**
Resolve https://github.com/JuliaReinforcementLearning/ReinforcementLearning.jl/issues/945

### CODE REVIEW

**Comment by CasBex** [comment]
I have checked this out and #945 is resolved with this. 

One small remark: it is not clear what `hook.empty_vec` is supposed to do. If it is private then maybe prefix it with an underscore?

**Comment by HenriDeh** [comment]
> what `hook.empty_vec` is supposed to do

Looks like some sort of cache. Could be renamed.

File: src/ReinforcementLearningCore/src/core/hooks.jl [hunk 1]
```diff
@@ -115,17 +114,21 @@ struct RewardsPerEpisode{T} <: AbstractHook where {T<:Number}
...
 end
 
 Base.getindex(h::RewardsPerEpisode) = h.rewards
 
-Base.push!(h::RewardsPerEpisode, ::PreEpisodeStage, agent, env) = push!(h.rewards, h.empty_vect)
-Base.push!(h::RewardsPerEpisode, ::PreEpisodeStage, agent, env, ::Symbol) = push!(h, PreEpisodeStage(), agent, env)
+function Base.push!(h::RewardsPerEpisode, ::PostEpisodeStage, agent, env)
+    push!(h.rewards, copy(h.empty_vect))
+    empty!(h.empty_vect)
+end
```
Comment (HenriDeh) [inline src/ReinforcementLearningCore/src/core/hooks.jl:126]:
```suggestion
function Base.push!(h::RewardsPerEpisode, ::PreEpisodeStage, agent, env)
    push!(h.rewards, Float64[])
end
```

File: src/ReinforcementLearningCore/src/core/hooks.jl [hunk 2]
```diff
@@ -115,17 +114,21 @@ struct RewardsPerEpisode{T} <: AbstractHook where {T<:Number}
...
+
+Base.push!(h::RewardsPerEpisode, ::PostEpisodeStage, agent, env, ::Symbol) = push!(h, PostEpisodeStage(), agent, env)
```
Comment (HenriDeh) [inline src/ReinforcementLearningCore/src/core/hooks.jl:128]:
```suggestion
Base.push!(h::RewardsPerEpisode, s::PreEpisodeStage, agent, env, ::Symbol) = push!(h, s, agent, env)
```

File: src/ReinforcementLearningCore/src/core/hooks.jl [hunk 3]
```diff
@@ -115,17 +114,21 @@ struct RewardsPerEpisode{T} <: AbstractHook where {T<:Number}
...
 
-Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(h.rewards[end], reward(env))
-Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E, player::Symbol) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(h.rewards[end], reward(env, player))
+Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(h.empty_vect, reward(env))
+Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E, player::Symbol) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(h.empty_vect, reward(env, player))
```
Comment (HenriDeh) [inline src/ReinforcementLearningCore/src/core/hooks.jl:131]:
```suggestion
Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(last(h.rewards), reward(env))
Base.push!(h::RewardsPerEpisode, ::PostActStage, agent::P, env::E, player::Symbol) where {P <: AbstractPolicy, E <: AbstractEnv} = push!(last(h.rewards), reward(env, player))
```

**Review by HenriDeh — COMMENTED** [review]
Can't we simplify this and directly work without this "empty_vec" ? Simply add a new empty vec directly to rewards at the start of episodes and push to that vec at each postact stage. 

We can then drop empty_vec from the struct.

**Comment by HenriDeh** [comment]
Otherwise I'm fine with approving right now. It's just that the bug reported by @CasBex showcases how a needlessly complex implementation easily leads to bugs.

**Comment by jeremiahpslewis** [comment]
> Otherwise I'm fine with approving right now. It's just that the bug reported by @CasBex showcases how a needlessly complex implementation easily leads to bugs.

Thanks for the fixes, they're right on!

**Review by HenriDeh — COMMENTED** [review]
Other than the tests not passing, we ought to drop the empty_vec field from the struct since it is not used anymore.

**Pull request merged**